src/_path.h

```diff
@@ -1089,35 +1089,39 @@ void quad2cubic(double x0, double y0,
 void __add_number(double val, char format_code, int precision,
                   std::string& buffer)
 {
-    char *str = PyOS_double_to_string(val, format_code, precision, 0, NULL);
-
-    // Delete trailing zeros and decimal point
-    char *q = str;
-    for (; *q != 0; ++q) {
+    if (precision == -1) {
+        // Special-case for compat with old ttconv code, which *truncated*
+        // values with a cast to int instead of rounding them as printf
+        // would do.  The only point where non-integer values arise is from
+        // quad2cubic conversion (as we already perform a first truncation
+        // on Python's side), which can introduce additional floating point
+        // error (by adding 2/3 delta-x and then 1/3 delta-x), so compensate by
+        // first rounding to the closest 1/3 and then truncating.
+        char str[255];
+        PyOS_snprintf(str, 255, "%d", (int)(round(val * 3)) / 3);
+        buffer += str;
+    } else {
+        char *str = PyOS_double_to_string(val, format_code, precision, 0, NULL);
+        // Delete trailing zeros and decimal point
+        char *q = str;
         // Find the end of the string
-    }
-
-    --q;
-    for (; q >= str && *q == '0'; --q) {
-        // Rewind through all the zeros
-    }
-
-    // If the end is a decimal point, delete that too
-    if (q >= str && *q == '.') {
+        while (*q) { ++q; }
         --q;
-    }
-
-    // Truncate the string
-    ++q;
-    *q = 0;
-
-    try {
-        buffer += str;
-    } catch (std::bad_alloc& e) {
+        // Rewind through all the zeros
+        while (q >= str && *q == '0') { --q; }
+        // If the end is a decimal point, delete that too
+        if (q >= str && *q == '.') { --q; }
```

Comment:
Stylistically, this might be against PEP7.

Reply:
I know, but I think splitting things on their own lines (as was done before) makes the code really harder to follow because the information density is so low...

Reply:
I went for a slight variation which does respect pep7.